src/main/java/net/fabricmc/loom/task/DownloadTask.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2025 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.task;
+
+import java.net.URISyntaxException;
+import java.time.Duration;
+
+import javax.inject.Inject;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.workers.WorkAction;
+import org.gradle.workers.WorkParameters;
+import org.gradle.workers.WorkQueue;
+import org.gradle.workers.WorkerExecutor;
+import org.jetbrains.annotations.ApiStatus;
+
+import net.fabricmc.loom.util.ExceptionUtil;
+import net.fabricmc.loom.util.download.Download;
+import net.fabricmc.loom.util.download.DownloadBuilder;
+import net.fabricmc.loom.util.download.DownloadException;
+
+/**
+ * A general purpose task for downloading files from a URL, using the loom {@link Download} utility.
+ */
+public abstract class DownloadTask extends DefaultTask {
+	/**
+	 * The URL to download the file from.
+	 */
+	@Input
+	public abstract Property<String> getUrl();
+
+	/**
+	 * The expected SHA-1 hash of the downloaded file.
+	 */
+	@Optional
+	@Input
+	public abstract Property<String> getSha1();
+
+	/**
+	 * The maximum age of the downloaded file in days. When not provided the downloaded file will never be considered stale.
+	 */
+	@Optional
+	@Input
+	public abstract Property<Integer> getMaxAge();
```

Comment:
Duration implements Serializable so you could probably expose them too (would allow for finer control)

Reply:
Ah nice, ill change that. I should have checked before assuming.

Reply:
Looks good, also added a basic test for it. It doesnt check it works (the unit tests do that), just makes sure gradle is happy with it.